src/tactic/polyrith.lean

```diff
@@ -0,0 +1,620 @@
+| n (_,m) := unify `(n) `(m)
+
+/--
+This can convert a `poly` into a `pexpr` that would evaluate to a polynomial.
+To do so, it uses an `exmap` that contains information about what atomic expressions
+the variables in the `poly` refer to.
+
+`poly` objects only contain coefficients from `ℚ`. However, the `poly` object might
+be referring to a polynomial over some other field. As such, the resulting `pexpr` contains
+no typing information.
+-/
+meta def poly.to_pexpr :exmap → poly → tactic pexpr
+| _ (poly.const z) := return z.to_pexpr
+| m (poly.var n) :=
+  do
+    (e, num) ← m.mfind (has_val n),
+    return ``(%%e)
+| m (poly.add p q) :=
+  do
+    p_pexpr ← poly.to_pexpr m p,
+    q_pexpr ← poly.to_pexpr m q,
+    return ``(%%p_pexpr + %%q_pexpr)
+| m (poly.sub p q) :=
+  do
+    p_pexpr ← poly.to_pexpr m p,
+    q_pexpr ← poly.to_pexpr m q,
+    if p_pexpr = ``(0) then return ``(- %%q_pexpr) else
+    return ``(%%p_pexpr - %%q_pexpr)
+| m (poly.mul p q) :=
+  do
+    p_pexpr ← poly.to_pexpr m p,
+    q_pexpr ← poly.to_pexpr m q,
+    return ``(%%p_pexpr * %%q_pexpr)
+| m (poly.pow p n) :=
+  do
+    p_pexpr ← poly.to_pexpr m p,
+    return ``(%%p_pexpr ^ %%n)
+
+
+/-!
+# Parsing SageMath output into a poly
+
+The following section contains code that can convert a string of appropriate format into
+a `poly` object. This is used later on to convert the coefficients given by Sage into
+`poly` objects.
+-/
+
+/--
+An error message returned from Sage has a kind and a message.
+-/
+@[derive inhabited]
+structure sage_error :=
+(kind : string)
+(message : string)
+
+open parser
+
+/--
+Parse an error message from the output of the polyrith python script.
+The script formats errors as "%{kind}%{message}".
+-/
+meta def error_parser : parser sage_error := do
+  ch '%',
+  kind ← many_char1 (sat (≠ '%')),
+  ch '%',
+  msg ← list.as_string <$> many any_char,
+  return ⟨kind, msg⟩
+
+/--
+A parser object that parses `string`s of the form `"poly.var n"`
+to the appropriate `poly` object representing a variable.
+Here, `n` is a natural number
+-/
+meta def var_parser : parser poly := do
+  str "poly.var ",
+  n ← nat,
+  return (poly.var n)
+
+/--
+A parser object that parses `string`s of the form `"-n"`
+to the appropriate `poly` object representing a negative integer.
+Here, `n` is a natural number
+-/
+meta def neg_nat_parser : parser int := do
+  ch '-',
+  n ← nat,
+  return (- n)
+
+/--
+A parser object that parses `string`s of the form `"-n"`
+to the appropriate `poly` object representing a positive integer.
+Here, `n` is a natural number
+-/
+meta def nat_as_int_parser : parser int := do
+  n ← nat,
+  return (n)
+
+/--
+A parser object that parses `string`s of the form `"poly.const z"`
+to the appropriate `poly` object representing a rational coefficient.
+Here, `n` is a rational number
+-/
+meta def const_fraction_parser : parser poly := do
+  str "poly.const ",
+  numer ← neg_nat_parser <|> nat_as_int_parser,
```

Comment:
`neg_nat_parser <|> nat_as_int_parser` should be extracted to a `parser.int` definition to match `parser.nat`. Similarly, there's a `parser.rat` definition hiding in here.